### ISSUE TITLE
Support for multi vnc

### ISSUE THREAD

**Issue by matthewei** [kubevirt/kubevirt#10553]
**What happened**:
I use this solution to realize [VNC](https://kubevirt.io/2019/Access-Virtual-Machines-graphic-console-using-noVNC.html). There is only one activing url when I open more than one url for the same vm 

**What you expected to happen**:
I can open more than one vnc-url for the same vm

**How to reproduce it (as minimally and precisely as possible)**:
Please follow the [VNC](https://kubevirt.io/2019/Access-Virtual-Machines-graphic-console-using-noVNC.html) and you can get the same result.

**Comment by dhiller** [comment]
@matthewei do IUC that you want to work on this? Assigning you just in case.

/assign @matthewei

**Comment by xpivarc** [comment]
This doesn't seem to do anything.

**Comment by matthewei** [comment]
> This doesn't seem to do anything.
yes, I realize it by changing a few codes. When one connection comes from api and the handler raise a goroutine to serve it.  So have verified it? It works well in my envionment.

**Comment by matthewei** [comment]
@dhiller Hi ,experts, Do you have any other comments for it?

**Comment by matthewei** [comment]
@0xFelix, sorry to bother you. 
I found that we could just open one vnc for a vm. when i analyzed the code i found that 
```go
func newStopChan(uid types.UID, lock *sync.Mutex, stopChans map[types.UID](chan struct{})) chan struct{} {
	lock.Lock()
	defer lock.Unlock()
	// close current connection, if exists
	if c, ok := stopChans[uid]; ok {
		delete(stopChans, uid)
		close(c)
	}
	// create a stop channel for the new connection
	stopCh := make(chan struct{})
	stopChans[uid] = stopCh
	return stopCh
}
you can see `close current connection, if exists`. So this is the root cause for that. In my opinion, it is necessay for a vm to create more than one vnc. So I will realize it.

**Comment by 0xFelix** [comment]
Okay, understood. Not sure what direction we should go here. Is there a reason to only allow a single VNC connection?

**Comment by dhiller** [comment]
Maybe because this might open for DDoS? In order to prohibit that we could have an upper boundary of maximum opwn connections. WDYT?

**Comment by xpivarc** [comment]
> Maybe because this might open for DDoS? In order to prohibit that we could have an upper boundary of maximum opwn connections. WDYT?

It is the decision of Kubevirt to not allow it, mainly because we tunnel the traffic through the Kubernetes API Server and our Server as well. This is not really scalable or in other words, can be a problem on a scale. Other than that I am not sure what policy we use for the VNC (shared or not).

**Comment by matthewei** [comment]
So we can limit the amount of vnc rather than to allow one vnc for customer.

**Comment by matthewei** [comment]
> Maybe because this might open for DDoS? In order to prohibit that we could have an upper boundary of maximum opwn connections. WDYT?也许是因为这可能会对 DDoS 开放？为了禁止我们可以有最大 opwn 连接的上限。WDYT？

sorry, I have no idea about it. I just feel confuesed about it. The VNC is just a websocket.

**Comment by 0xFelix** [comment]
We don't want to flood the control plane with port forwardings resp. websockets in general. This is a way to quickly gain access to a VM's VNC (or other ports) for troubleshooting purposes. If you want to expose the VNC console of VM's at scale you should probably have a look at services. WDYT? @matthewei @xpivarc

**Comment by matthewei** [comment]
> We don't want to flood the control plane with port forwardings resp. websockets in general. This is a way to quickly gain access to a VM's VNC (or other ports) for troubleshooting purposes. If you want to expose the VNC console of VM's at scale you should probably have a look at services. WDYT? @matthewei @xpivarc我们不想用端口转发或一般的 websocket 淹没控制平面。这是一种快速访问 VM 的 VNC（或其他端口）以进行故障排除的方法。如果要大规模公开 VM 的 VNC 控制台，则可能应该查看服务。WDYT？@matthewei @xpivarc

I see. The openstack's soluation uses Traffic-network to realize it. Actually, we do need multi-vnc. I want to realize it just as openstack. The guest who want to use vnc can get a public-url which is similier to openstack. Do you have another idea?

**Comment by 0xFelix** [comment]
As mentioned, you could explore Services. They allow you to make ports on a VM publicly available. This way restrictions on concurrent connections by KubeVirt do no longer apply.

https://kubernetes.io/docs/concepts/services-networking/service/

**Comment by victortoso** [comment]
I'm not one to go against VDI in KubeVirt ([example](https://github.com/kubevirt/kubevirt/pull/10971)) but in this case, I agree with what [was said](https://github.com/kubevirt/kubevirt/issues/10553#issuecomment-1812984049) that we should be careful in overusing the websocket subresource... 

The more long term solution would be to have vnc or rdp server to run outside of QEMU process. Upstream this is already doable [at least with RDP](https://crates.io/crates/qemu-rdp)... and then some discussion on how to integrate that with KubeVirt, hopefully not relying on virt-handler to bridge the connection.

**Comment by chinglinwen** [comment]
Thanks for the interesting link of https://crates.io/crates/qemu-rdp. 

I understand your concern, and agree to close this pull request. I'll using this fix for internal fork(for quick fix), if later KubeVirt implemented vnc or rdp server solution, we than just do the switch.
